Support create table if not exists

taeyeon-Kim opened the pull request:
## Impact

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes expected existing functionality)
- [x] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Description

Support create table if not exists to DatabaseChangeLogTable.
If a spring test context has a `@MockBean` or `@SpyBean`, it creates a new context rather than using the cached context.
At this time, the "Already exists ~" error occurs when the DatabaseChangeLog table is created once more as the liquibase bean is newly created. 

## Things to be aware of

- Describe the technical choices you made
- Describe impacts on the codebase

## Things to worry about

- List any questions or concerns you have with the change
- List unknowns you have 

## Additional Context

Add any other context about the problem here.

taeyeon-Kim commented:
+1

nvoxland commented:
This PR makes too many assumptions to be able to take in as-is. CreateTableGenerator assumes that it can use `IF NOT EXISTS` if the flag is set, but not all (maybe not even most?) databases support that. 

You could add validation logic to the CreateTableGenerator to fail if the flag is not supported for a given database, but then you have to deal with it on `CreateDatabaseChangeLogTableGenerator.java` and only set that flag if the statement is supported. 

Otherwise, this will cause liquibase to break on a lot of databases since the creation of the databasechangelog table is so central. I set the build to run to get an idea of where the test failures will be of our curiosity.

Are you able to fix up the PR to be less breaking to other users, @taeyeon-Kim? Or is there a different approach that would address your problem better?

taeyeon-Kim commented:
@nvoxland 
Most cases like mine are caused by the life cycle difference between Inmemory DB (like H2) and app context.
So I fixed it to apply "if not exist" SQL only on H2.

nvoxland commented:
Thanks @taeyeon-Kim. Although rather than just silently ignoring the flag for other databases, it's better to add it to the `validate` method in `CreateTableGenerator` to fail if it's set on databases that don't support it. For now, you could have that just for h2 although I'm pretty sure it work for postgresql and mysql at least too.

taeyeon-Kim commented:
@nvoxland 
Thank you for considering the new approach positively.
I created `supportIfNotExists()` method in the `Database` interface and checked other DB.
I checked the DB supported "if not exists" by Liquidibase, is it possible to double check it?

### Support
- Cockroach: https://www.cockroachlabs.com/docs/stable/create-table.html#parameters
- DB2: https://www.ibm.com/docs/en/db2-warehouse?topic=statements-create-table
- H2: https://www.h2database.com/html/commands.html#create_table
- Hsql (uncertain...)
   - http://www.hsqldb.org/doc/1.8/guide/ch09.html#create_table-section
   - https://www.hsqldb.org/doc/2.0/guide/databaseobjects-chapt.html
- Informix: https://www.ibm.com/docs/en/informix-servers/14.10?topic=statements-create-table-statement
- Ingres9: https://docs.actian.com/ingres/11.2/index.html#page/SQLRef/CREATE_TABLE.htm
- mariaDB: https://mariadb.com/kb/en/create-table/#create-table-if-not-exists
- postgres: https://www.postgresql.org/docs/current/sql-createtable.html
- SQLite: https://www.sqlite.org/lang_createtable.html

### unsupport
 - Db2z: https://www.ibm.com/docs/en/db2-for-zos/12?topic=statements-create-table
 - Derby: https://db.apache.org/derby/docs/10.16/ref/rrefsqlj24513.html
 - EnterpriseDB: https://www.enterprisedb.com/docs/epas/latest/epas_compat_sql/36_create_table/
 - Firebird: https://firebirdsql.org/refdocs/langrefupd15-create-table.html
 - MSSQL: https://learn.microsoft.com/en-us/sql/t-sql/statements/create-table-transact-sql?view=sql-server-ver16
 - Oracle: https://docs.oracle.com/database/121/SQLRF/statements_7002.htm#SQLRF01402
 - Sybase: https://infocenter.sybase.com/help/index.jsp?topic=/com.sybase.infocenter.dc00801.1510/html/iqrefso/X315712.htm

taeyeon-Kim commented:
@nvoxland Please check the comments :)

taeyeon-Kim commented:
@nvoxland Please check the comments :)

taeyeon-Kim commented:
@nvoxland Please check the comments :)

nvoxland commented:
Thanks for the research on all those!

The new `supportIfNotExists` method on the interface does need a default implementation (of "false") so it's not a breaking API change for any extensions that implement Database but haven't overridden that method.

I also think the method should be a bit more explicit in the name as `supportCreateIfNotExists` and should have a parameter of `Class<DatabaseObject> type` so it can someday be used for procedures or anything else, using `if (database.supportsCreateIfNotExists(StoredProcedure.class)`. Your implementations could still just check for tables for now, but it keeps the API more flexible for future use.

taeyeon-Kim commented:
@nvoxland Thank you for the idea
Are you asking me to add the default method to the interface?
Other methods in the `Database` interface also matched the convention without using the default method.
So the default was implemented in `AbstractJdbcDatabase`.
also changed the signature to `supportCreateIfNotExists(Class<?extends DatabaseObject> type)`.

taeyeon-Kim commented:
@nvoxland Thank you for the idea
Are you asking me to add the default method to the interface?
Other methods in the `Database` interface also matched the convention without using the default method.
So the default was implemented in `AbstractJdbcDatabase`.
also changed the signature to `supportCreateIfNotExists(Class<?extends DatabaseObject> type)`.

taeyeon-Kim commented:
@nvoxland Please check the comments :)

taeyeon-Kim commented:
@nvoxland Please check the comments :)

filipelautert commented on liquibase-core/src/main/java/liquibase/database/Database.java:
```diff
@@ -472,5 +472,7 @@ default String getDisplayName() {
 
     ValidationErrors validate();
+
+    boolean supportsCreateIfNotExists(Class<? extends DatabaseObject> type);
```
Great, this way we can validate if it supports any kind of "exists" syntax. Could you please add simple javadocs explaining this method?

filipelautert commented on liquibase-core/src/main/java/liquibase/database/core/InformixDatabase.java:
```diff
@@ -26,85 +27,85 @@ public class InformixDatabase extends AbstractJdbcDatabase {
         super.setCurrentDateTimeFunction("CURRENT " + DATETIME_FIELD_QUALIFIER);
         super.sequenceNextValueFunction = "%s.NEXTVAL";
         systemTablesAndViews.add("systables");
-		systemTablesAndViews.add("syscolumns");
```
Whow, seems this class really needed some identation fix :open_mouth:  Thanks

filipelautert commented on liquibase-core/src/main/java/liquibase/sqlgenerator/core/CreateTableGenerator.java:
```diff
@@ -58,7 +59,13 @@ public Sql[] generateSql(CreateTableStatement statement, Database database, SqlG
-        buffer.append("TABLE ").append(database.escapeTableName(statement.getCatalogName(),
+        buffer.append("TABLE ");
+
+        if (database.supportsCreateIfNotExists(StoredProcedure.class)) {
```
Hey, StoredProcedure was an example that Nathan used as another ways that we can use this method. So here you should be passing Table.class as a parameter (and then change all the other classes that are testing against it). So if in the future we change the stored procedure logic to add "if exists", then we will add it ;)

filipelautert replied:
Thanks!

filipelautert reviewed (commented):
Hello @taeyeon-Kim ! After Nathan pre-review I'm running a second pass. Can you please see my review comments ? 
I also noticed that it broke almost all tests on CreateTableGeneratorTest . Maybe it will need a litle refactoring: instead of just using the string "CREATE TABLE " perhaps it will need a method or something like this to generate the string appending the "if exists" if the database supports it.

taeyeon-Kim commented:
@filipelautert Thanks for the review.
- Add javadoc to the `supportsCreateIfNotExists` method.
- fix some broken tests.
- fix to pass a Table, not a StoredProcedure, as a parameter.

taeyeon-Kim commented:
@filipelautert I rebased it after conflicts.

filipelautert commented:
Hello @taeyeon-Kim
Changes on core are fine, but now the test projects are breaking because of this PR - as it changes one of the core SQLs for liquibase, a lot of tests are failing. 
I started fixing test-harness project on PR https://github.com/liquibase/liquibase-test-harness/pull/500 , and after that I need to fix integration tests too. I'll be working on that on the next days, thanks for your help so far!

taeyeon-Kim commented:
@filipelautert This is an issue not only in testing, but also when using springboot and h2 file mode.
It's because I'm trying to create liquibase related tables in the database... I know you're busy, but this issue seems to happen everywhere with springboot.

taeyeon-Kim commented:
@filipelautert 
Ah... The issue was caused by using the `DATABASE_TO_LOWER=TRUE;` option in h2.
After deleting that option, the "Already exists ~" case disappeared and it works fine.
If you think this PR will help in other cases besides this one, I guess we can go ahead with it.

filipelautert commented:
Hey @taeyeon-Kim , great that you found a workaround.
Still this PR will help and we plan to carry it ahead, we just need to release next week and then I can concentrate on this one againn a fix all the failing tests.

filipelautert commented:
Hello @taeyeon-Kim - Nathan and I discussed this PR and we think that it could cause some issues to the way that people are used to liquibase by changing the default behavior of create table. So maybe instead of "if exists" be appended by default, it could be defined as a parameter to "createTable" tag. I believe this would not resolve your initial problem, but now that it is resolved we could change the PR to behave this way. What do you think?

taeyeon-Kim commented:
@filipelautert That's a good idea. I was worried that this PR was a change to liquibase core, which is why I kept sharing the situation in the comments. If you could share a little more detail about your new idea, I'd be happy to edit it.
Is that okay?

filipelautert commented:
@taeyeon-Kim I think that PR https://github.com/liquibase/liquibase/pull/3108 is a good example: it introduces a new attribute on createTable and changes the create table expression if enabled. 
So having the mix of a new attribute + your validation of which databases support "if exists" would be perfect! Thank you for working on this.

taeyeon-Kim commented:
@filipelautert I've got a new commit to fix the conflict and rebase...  Should I just go with the commit from 6d5ba4e1b63e0f883cf22d6522d96a1cc9c424b0 ?
Of course, this isn't finished... I just wanted to see if this idea works.

filipelautert commented:
I think it's good! 
We also just need to revert all of my commits that were fixing the tests...
Maybe it's easier to go back to commit 9e3c27e574b4d74f5bc064a35372f88d207c674c , apply the StoredProcedure by Table replacements (that you added on 3f9be3ebf0eecafae45691a7808e1a6a87e8c7d6 but it also has some test fixes) , cherry-pick you latest commit and force push it?
Or just revert my commits and the tests changes from 3f9be3ebf0eecafae45691a7808e1a6a87e8c7d6  , whichever you think it's better :)

taeyeon-Kim commented:
@filipelautert I've cleaned up the commits and added functionality
However, I have 2 questions.
1. how do we enable this feature for users, or is this just a temporary feature that we can disable if there is no way?

2. isn't PR #3108 feature not working?? The code below is always empty...
```java
if (statement.getTableType() != null && statement.getTableType().trim().isEmpty()) {
            buffer.append(statement.getTableType().toUpperCase());
        }

```

filipelautert commented on liquibase-core/src/main/resources/www.liquibase.org/xml/ns/dbchangelog/dbchangelog-latest.xsd:
```diff
@@ -1004,6 +1004,7 @@
             <xsd:attribute name="tablespace" type="xsd:string"/>
             <xsd:attribute name="tableType" type="xsd:string"/>
             <xsd:attribute name="remarks" type="xsd:string"/>
+            <xsd:attribute name="enableIfNotExistsStatement" type="xsd:boolean"/>
```
I believe the attribute name could be shorter, like "ifNotExists" ?

taeyeon-Kim replied:
@filipelautert fixed it in ec46d1dc271163ea395cfb0a60ed6f1822f7757e

filipelautert commented:
Hello @taeyeon-Kim ! 

1. That's an "optional" feature, so the users can enable it on their changesets like:
```
<createTable tableName="book" ifNotExists="true">
....
```
2.  You're damn right, it's not working, thanks for noticing that 🤣 It should be something like :
```
        if (StringUtil.isNotEmpty(statement.getTableType())) {
            buffer.append(statement.getTableType().toUpperCase());
        }
```

taeyeon-Kim commented:
@filipelautert Reviewed by ec46d1dc271163ea395cfb0a60ed6f1822f7757e. 
I made sure that the CreateTableGeneratorInformix class, which extends CreateTableGenerator for Informix only, also reflects this.

filipelautert commented:
Hello @taeyeon-Kim - Yesterday I fixed some tests and did some small changes to make it all work together. I think the PR is good to go! Do you want to do any last changes or is it fine to you too?

taeyeon-Kim commented:
@filipelautert No. Thank you for your help in making this feature a reality.

filipelautert reviewed (approved):
This pr adds a new "ifnotexists" properties to create table, and if it is set to true and the database supports it the create table expression is changed to "CREATE TABLE IF NOT EXISTS..."

Thanks @taeyeon-Kim for the code!

suryaaki2 commented:
This is a great enhancement, thanks @taeyeon-Kim! Should we change this PR to enhancement from bug fix?
@filipelautert

taeyeon-Kim commented:
Oh, there was a conflict, I resolved it and rebased.

taeyeon-Kim commented:
@filipelautert Oh, there was a conflict, I resolved it and rebased.

filipelautert commented:
@XDelphiGrl - build and tests are ok now.

XDelphiGrl commented:
Rerunning test-harness SQL Server tests here: [https://github.com/liquibase/liquibase-test-harness/actions/runs/5260112315](https://github.com/liquibase/liquibase-test-harness/actions/runs/5260112315)

XDelphiGrl commented on liquibase-standard/src/test/groovy/liquibase/change/core/CreateTableChangeTest.groovy:
```diff
@@ -324,4 +333,30 @@ public class CreateTableChangeTest extends StandardChangeTest {
+        SqlGeneratorFactory.getInstance().generateSql(change, database)[0].toSql() == expected
+
+        where:
+        type          | autoinc | database               | expected
```
I believe we should add test cases for the databases that do **not** support `CREATE TABLE ... IF NOT EXISTS` just to ensure that the `IF NOT EXISTS` does not get added to the generated SQL unexpectedly.

CC @rberezen @filipelautert

XDelphiGrl reviewed (changes requested):
In the comments, I've requested an addition to the `CreateTableChangeTest.groovy`. 

CHANGES REQUESTED

rberezen commented:
@taeyeon-Kim I hope you're doing well. I wanted to check if you had the opportunity to review the changes requested by @XDelphiGrl

taeyeon-Kim commented:
@XDelphiGrl @rberezen Sorry for the delay, I was working on something else, I'll update :)

taeyeon-Kim commented on liquibase-standard/src/main/java/liquibase/database/core/EnterpriseDBDatabase.java:
```diff
@@ -58,4 +59,9 @@ public String getDefaultDriver(String url) {
     }
+
+    @Override
+    public boolean supportsCreateIfNotExists(Class<? extends DatabaseObject> type) {
```
EnterpriseDBDatabase doesn't support the if NOT EXISTS statement, but because it inherits from PostgresDatabase, it's set to true.
Fix this as well.

rberezen commented on liquibase-standard/src/main/java/liquibase/database/core/InformixDatabase.java:
```diff
@@ -141,78 +142,78 @@
-		return correct;
-	}
+            throws DatabaseException {
+        Boolean correct = false;
```
@taeyeon-Kim Would you mind taking a look at this potential problem?

taeyeon-Kim replied:
@rberezen Cleaned up in 37ea712dfd266cd80e7ed1b4526457a063eb2edf.

rberezen reviewed (approved):
@taeyeon-Kim, I deeply appreciate your dedication and effort in handling this pull request.
This PR introduces a new 'ifNotExists' attribute for createTable change type.

- Functional and test harness executions passing. There are only some known snowflake test failures. 
- New unit test added.
- Manual testing completed.
- Documentation ticket created [PD-3594](https://datical.atlassian.net/browse/PD-3594)
 
APPROVED

taeyeon-Kim commented:
Thank you, it was a great experience to contribute to liquibase.

adrian-velonis1 commented:
Thanks for this PR and really nice work. Adding a brief description in the metadata for docs.

taeyeon-Kim commented:
@adrian-velonis1 @rberezen @filipelautert Hello :)
Is there anything else I can do to add to the documentation or help you work on?

filipelautert commented:
@taeyeon-Kim sorry for the delay to get this merged. We need to resolve the merge conflicts then it can be finally documented merged.

taeyeon-Kim commented:
@filipelautert 
I resolved the conflict and rebased it.
Please check again to make sure I didn't do anything wrong.  😃

filipelautert commented:
@rberezen I did some updates on this PR to bring it in sync with our latest code. Do you mind giving it a last review so we can get it merged still this year? thanks!

filipelautert commented:
@taeyeon-Kim took a while but it's merged! Thanks for the pr and sorry for the long delay to get it merged.

Now the next step is to identify other if not exists (I think we have it at least for view) and migrate them to this new implementation 🚀

adrian-velonis1 commented:
Internal, ignore this: https://datical.atlassian.net/browse/PD-3594
